packages/server/core/src/base/middlewares/renderHandler/render.ts

```diff
@@ -104,21 +104,27 @@ export async function createRender({
       metrics,
     };
 
+    let response: Response | void;
+
     switch (renderMode) {
       case 'data':
-        // eslint-disable-next-line no-case-declarations
-        let response = await dataHandler(req, renderOptions);
+        response = await dataHandler(req, renderOptions);
         if (!response) {
           response = await renderHandler(req, renderOptions, 'ssr', onError);
         }
+        break;
 
-        return response;
       case 'ssr':
       case 'csr':
-        return renderHandler(req, renderOptions, renderMode, onError);
+        response = await renderHandler(req, renderOptions, renderMode, onError);
+        break;
       default:
         throw new Error(`Unknown render mode: ${renderMode}`);
     }
+
+    applyExtendHeaders(response, routeInfo);
```

Comment:
Should the data handler's response be extended with headers?

Reply:
done~